data-mongodb/src/main/java/io/micronaut/data/mongodb/operations/MongoUtils.java

```diff
@@ -118,6 +121,16 @@ static Object toValue(BsonValue bsonValue) {
                 return Instant.ofEpochMilli(bsonValue.asDateTime().getValue());
             case NULL:
                 return null;
+            case DOCUMENT:
+                BsonDocument bsonDocument = bsonValue.asDocument();
```

Comment:
Not 100% sure this is the right solution, but `BsonDocument` exposed keySet, entrySet etc. like map properties so it made sense to me and this way it fixes given issue.